File: src/VCFIsolated.cpp
#include <algorithm>
#include <cassert>
#include "../include/VCFIsolated.h"
#include "../include/OptionSmall.h"
#include "../include/common.h"

using namespace std;


//////////////////// VCFIsolated ////////////////////

VCFIsolated::VCFIsolated(const STRVEC& s, size_t nis, vector<GenoRecord *>& rs,
											const Map& m, const VCFSmall *vcf) :
									VCFClippable(s, m, vcf),
									records(rs), num_imputed_samples(nis) { }

VCFIsolated::~VCFIsolated() {
	for(auto p = records.begin(); p != records.end(); ++p)
		delete *p;
}

vector<VCFIsolated *> VCFIsolated::divide_by_cM() const {
	// If more than 1 cM but not more than 10 pieces and not more than 10 cM
	// are considered to be a block
	std::vector<VCFIsolated *>	vcfs;
	vector<GenoRecord *>	rs(1, records[0]);
	for(std::size_t i = 1; i < this->size(); ++i) {
		GenoRecord	*record = records[i];
		if(this->is_block(record, rs)) {
			rs.push_back(record);
		}
		else {
			vcfs.push_back(new VCFIsolated(get_samples(), num_imputed_samples,
												rs, get_map(), get_ref_vcf()));
			rs.clear();
			rs.push_back(record);
		}
	}
	vcfs.push_back(new VCFIsolated(get_samples(), num_imputed_samples,
												rs, get_map(), get_ref_vcf()));
	return vcfs;
}

vector<Haplotype> VCFIsolated::collect_haplotype_from_refs() const {
	vector<Haplotype>	haps;
	for(size_t i = num_imputed_samples; i < num_samples(); ++i) {
		haps.push_back(clip_haplotype(i, 0));
		haps.push_back(clip_haplotype(i, 1));
	}
	return haps;
}

vector<Haplotype> VCFIsolated::collect_haplotypes_mat(
										size_t sample_index) const {
	return collect_haplotype_from_refs();
}

vector<Haplotype> VCFIsolated::collect_haplotypes_pat(
										size_t sample_index) const {
	return collect_haplotype_from_refs();
}

vector<HaplotypePair> VCFIsolated::impute_cM(
									const vector<HaplotypePair>& prev_haps) {
	vector<HaplotypePair>	haps;
	for(size_t i = 0; i < prev_haps.size(); ++i) {
		const auto	prev_hap = prev_haps[i];
		const auto	hap = this->impute_cM_each_sample(prev_hap, i, true);
		haps.push_back(hap);
	}
	return haps;
}

void VCFIsolated::impute() {
	const Haplotype	h = Haplotype::default_value();
	vector<HaplotypePair>	haps(num_imputed_samples, make_pair(h, h));
	vector<VCFIsolated *>	vcf_cMs = this->divide_by_cM();
	for(auto p = vcf_cMs.begin(); p != vcf_cMs.end(); ++p) {
		VCFIsolated	*vcf_cM = *p;
		haps = vcf_cM->impute_cM(haps);
		vcf_cM->records.clear();
		delete vcf_cM;
	}
	cout << num_imputed_samples << " samples are imputed." << endl;
}

VCFIsolated *VCFIsolated::create(const VCFSmall *orig_vcf,
											const VCFGeno *imputed_vcf,
											const STRVEC& samples,
											const STRVEC& references,
											const OptionSmall& op) {
	const auto	columns = orig_vcf->extract_columns(samples);
	const auto	ref_columns = imputed_vcf->extract_columns(references);
	STRVEC	new_samples;
	for(auto p = columns.begin(); p != columns.end(); ++p) {
		new_samples.push_back(orig_vcf->get_samples()[*p-9]);
	}
	new_samples.insert(new_samples.end(), references.begin(), references.end());
	vector<GenoRecord *>	records;
	for(size_t i = 0; i < orig_vcf->size(); ++i) {
		VCFRecord	*record = orig_vcf->get_record(i);
		GenoRecord	*imputed_record = imputed_vcf->get_record(i);
		const ll	pos = imputed_record->get_pos();
		vector<int>	geno;
		for(auto p = columns.begin(); p != columns.end(); ++p)
			geno.push_back(Genotype::all_gt_to_int(record->get_v()[*p]));
		for(auto q = ref_columns.begin(); q != ref_columns.end(); ++q)
			geno.push_back(imputed_record->get_geno()[*q]);
		GenoRecord	*new_record = new GenoRecord(pos, geno);
		records.push_back(new_record);
	}
	return new VCFIsolated(new_samples, samples.size(),
									records, op.map, orig_vcf);
}

File: python/VCFIsolated.py
from __future__ import annotations

# coding: utf-8
# VCFIsolated.py
# isolated samples

from collections import defaultdict, Counter
from typing import List, Tuple, Optional, IO, Dict, Iterator

from VCF import VCFSmall
from VCFGeno import VCFGenoBase, VCFGeno
from GenoRecord import GenoRecord
from Haplotype import *
import RefCommon
from Genotype import Genotype
from OptionSmall import *


#################### VCFIsolated ####################

# imputed samples at the beginning, followed by reference samples
class VCFIsolated(VCFGenoBase, VCFMeasurable):
	def __init__(self, samples: list[str], num_imputed_samples: int,
						records: list[GenoRecord], map_: Map, vcf: VCFSmall):
		VCFGenoBase.__init__(self, samples, vcf)
		VCFMeasurable.__init__(self, map_)
		self.records: list[GenoRecord] = records
		self.num_imputed_samples = num_imputed_samples
	
	##### virtual methods for VCFGenoBase #####
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> GenoRecord:
		return self.records[i]
	
	def get_records(self) -> list[GenoRecord]:
		return [ r for r in self.records ]
	
	##### non-virtual methods #####
	# divide VCF by 1 cM
	def divide_by_cM(self) -> Iterator[VCFIsolated]:
		# 1cMを超えても10個以内で10cM以内なら塊とみなす
		rs = self.records[:1]
		for i in range(1, len(self)):
			record = self.records[i]
			if self.is_block(record, rs):
				rs.append(record)
			else:
				yield VCFIsolated(self.samples, self.num_imputed_samples,
														rs, self.map, self.vcf)
				rs = [record]
		
		yield VCFIsolated(self.samples, self.num_imputed_samples,
														rs, self.map, self.vcf)
	
	def collect_haplotype_from_refs(self) -> list[Haplotype]:
		return [ self.clip_haplotype(i, j)
					for i in range(self.num_imputed_samples, self.num_samples())
					for j in (0, 1) ]
	
	def collect_haplotypes_mat(self, sample_index: int) -> list[Haplotype]:
		return self.collect_haplotype_from_refs()
	
	def collect_haplotypes_pat(self, sample_index: int) -> list[Haplotype]:
		return self.collect_haplotype_from_refs()
	
	def clip_haplotype(self, sample_index: int, side: int) -> Haplotype:
		hap = self.clip_raw_haplotype(sample_index, side)
		return Haplotype(hap, sample_index, side)
	
	def is_block(self, record: GenoRecord, rs: list[GenoRecord]) -> bool:
		length = self.cM(record.pos) - self.cM(rs[0].pos)
		if length < 1.0:
			return True
		elif len(rs) < 10 and length < 10.0:
			return True
		else:
			return False
	
	def get_int_gts(self, sample_index: int) -> list[int]:
		int_gts: list[int] = [ self.get_record(i).unphased(sample_index)
												for i in range(len(self)) ]
		return int_gts
	
	def impute_cM_each_sample(self, prev_hap: HaplotypePair,
								sample_index: int, exec: bool) -> HaplotypePair:
		int_gts: list[int] = self.get_int_gts(sample_index)
		haps_mat: list[Haplotype] = self.collect_haplotypes_mat(sample_index)
		haps_pat: list[Haplotype] = self.collect_haplotypes_pat(sample_index)
		seed: int = self.get_record(0).pos
		hap = Haplotype.impute(int_gts, haps_mat, haps_pat, prev_hap, seed)
		if exec:	# actually impute?
			self.set_haplotype(hap, sample_index)
		return hap
	
	def set_haplotype(self, hap: HaplotypePair, sample_index: int) -> None:
		hap_mat, hap_pat = hap
		for i in range(len(self)):
			record = self.get_record(i)
			gt = hap_mat.hap[i] | (hap_pat.hap[i] << 1) | 4
			record.geno[sample_index] = gt
	
	def impute_cM(self, prev_haps: list[HaplotypePair]) -> list[HaplotypePair]:
		haps: list[HaplotypePair] = []
		for sample_index, prev_hap in enumerate(prev_haps):
			hap = self.impute_cM_each_sample(prev_hap, sample_index, True)
			haps.append(hap)
		return haps
	
	def impute(self) -> None:
		h = Haplotype.default_value()
		haps = [(h, h)] * self.num_imputed_samples
		for vcf_cM in self.divide_by_cM():
			haps = vcf_cM.impute_cM(haps)
		print("%d samples are imputed." % self.num_imputed_samples)
	
	@staticmethod
	def create(orig_vcf: VCFSmall, merged_vcf: VCFGeno,
						samples: list[str], references: list[str],
						op: OptionSmall) -> VCFIsolated:
		samples.sort()	# C++と合わせるため
		columns = orig_vcf.extract_columns(samples)
		ref_columns = merged_vcf.extract_columns(references)
		new_samples = [ orig_vcf.samples[c-9] for c in columns ] + references
		header = orig_vcf.trim_header(new_samples)
		new_records: list[GenoRecord] = []
		for i in range(len(orig_vcf)):
			record = orig_vcf.records[i]
			imputed_record = merged_vcf.records[i]
			pos = imputed_record.pos
			geno = ([ Genotype.all_gt_to_int(record.v[c]) for c in columns ] +
								[ imputed_record.geno[c] for c in ref_columns ])
			new_record = GenoRecord(pos, geno)
			new_records.append(new_record)
		vcf = VCFIsolated(new_samples, len(samples),
								new_records, op.map, orig_vcf)
		return vcf
	
	@staticmethod
	def create_ref(orig_vcf: VCFSmall, phased_vcf: VCFGeno,
						samples: list[str], references: list[str],
						op: OptionSmall) -> VCFIsolated:
		vcf = VCFGeno.extract_samples(samples, orig_vcf);
		new_samples = samples + references
		records = RefCommon.merge_records(phased_vcf, vcf, new_samples)
		return VCFIsolated(new_samples, len(samples),
									records, op.map, orig_vcf)

__all__ = ['VCFIsolated']
